### PR TITLE
Use API to initiate repository installation.

### PR DESCRIPTION
**Pull request by davebx**

### CODE REVIEW

**Comment by martenson** [comment]
I like this change! Installation tested locally.Thanks for working on this
:+1:

**Comment by bgruening** [comment]
I'm just wondering if there is a better name as `initiate_repository_installation`. Can we kill `initiate_`?

**Comment by davebx** [comment]
I'm open to making it `repository_installation`

**Comment by martenson** [comment]
`install_repository` ?

**Comment by bgruening** [comment]
:+1: for  `install_repository`

**Comment by dannon** [comment]
Or even just `install`?  I mean, we're already at /api/tool_shed_repositories/, right?

**Comment by davebx** [comment]
:+1: for `install`

**Comment by nsoranzo** [comment]
I think it's not very clear the difference between `initiate_repository_installation()` and `install_repository_revision()`.

**Comment by dannon** [comment]
@nsoranzo What is the difference?  The docstrings don't make that clear to me either.

File: templates/admin/tool_shed_repository/initiate_repository_installation.mako [hunk 1]
```diff
@@ -17,11 +17,11 @@
             // Make ajax call
             $.ajax( {
                 type: "POST",
-                url: "${h.url_for( controller='admin_toolshed', action='manage_repositories' )}",
+                url: "/api/tool_shed_repositories/initiate_repository_installation",
```
Comment (jmchilton) [inline templates/admin/tool_shed_repository/initiate_repository_installation.mako]: Will this work if Galaxy is served on a prefix - I feel like this was using url_for a reason.

**Comment by davebx** [comment]
@nsoranzo @dannon `install_repository_revision(s)` both wait until the installations complete before returning status, whereas `install` just tells the galaxy process to start installing, while the installation monitoring page updates the status.

File: templates/admin/tool_shed_repository/initiate_repository_installation.mako (hunk 1 above)
Comment (davebx) [inline templates/admin/tool_shed_repository/initiate_repository_installation.mako]: This is a good point. Someone should fix that.

**Comment by dannon** [comment]
Ahh.  Can we just unify this instead of having two endpoints then, perhaps with an optional 'block' flag or something?

(though I'm having trouble seeing when blocking is a really good idea wrt: web requests like this that might take a while)

**Comment by davebx** [comment]
@dannon Yes. As for the blocking, that API method was designed for command line installation of repositories.

**Comment by yhoogstrate** [comment]
This is a feature our group is/was waiting for. Great!

**Comment by martenson** [comment]
xref: https://trello.com/c/uCWVDCsX/19-toolshed-make-resetting-metadata-on-a-selected-list-of-repositories-in-both-galaxy-and-the-tool-shed-ajaxian

**Comment by bgruening** [comment]
I assume that one reason this was a blocking API is that it is not possible to install multiple repositories at the same time. What about adding a warning message about this into the API?

**Comment by dannon** [comment]
I'm not sure about the origin story here (and probably don't want to look ;)), but in any event a web request should basically _never_ block unless it's a < 500ms (and that's being _really_ generous) operation, API or not.

This should return a handle to an ongoing install process (which is really just the tool identifier, since everything else is derivable from there) and command line tools (and anything else) should poll for status.

**Comment by davebx** [comment]
@dannon I like that idea. And then the endpoint for installation status updates could potentially be something like `/repositories/install/{status identifier}/status`

Although I'm not sure the tool identifier, unless it's a GUID, would be specific enough. I'll give it some thought, maybe bother @martenson about it, and get back to you.

**Comment by dannon** [comment]
@davebx Yep, I assumed the full tool guid.

**Comment by hexylena** [comment]
FYI @galaxybot will comment on PRs with Emoji +1s, so use text +1 if you want it to ignore your comments

**Comment by martenson** [comment]
@davebx can we get a rebase please?

**Comment by davebx** [comment]
This code will be included in my upcoming pull request based on https://github.com/davebx/galaxy/tree/toolshed_decouple

**Pull request closed without merging**